### PR TITLE
Add TVM backend

### PR DESCRIPTION
**Pull request by interesaaat**
Closes #236.

### CODE REVIEW

**Comment by interesaaat** [comment]
I was able to make some operator work (linears and some tree with GEMM; not all trees, like isolation forest throw a compilation error). Unfortunately other operators do not work out of the box. I opened a [PR](https://github.com/apache/incubator-tvm/issues/6282) regarding `index_select`. If support for `index_select` is added, I think we should have `tree_trav` and `perf_tree_trav` strategies working too.

File: hummingbird/ml/_topology.py [hunk 1]
```diff
@@ -185,6 +218,79 @@ def fix_graph(graph):
...
+            config["relay.FuseOps.max_depth"] = 50
+
+        # Create the relay version of the model.
+        model, params = relay.frontend.from_pytorch(ts_model, test_input)
+        if remainder_trace_input is not None:
+            remainder_model, remainder_params = relay.frontend.from_pytorch(remainder_ts_model, remainder_test_input)
+
+        # Generate the model.
+        with tvm.transform.PassContext(opt_level=3, config=config):
+            opt_mod, opt_params = relay.optimize(model, target=target, params=params)
```
Comment (masahi) [inline hummingbird/ml/_topology.py]: no need to call optimize explicitly, `relay.build` calls `optimize` inside

File: hummingbird/ml/_topology.py [hunk 2]
```diff
@@ -185,6 +218,79 @@ def fix_graph(graph):
...
+            graph, lib, params = relay.build(opt_mod, target=target, params=opt_params)
+        tvm_model = graph_runtime.create(graph, lib, ctx)
+        tvm_model.set_input(**params)
+        if remainder_trace_input is not None:
+            with tvm.transform.PassContext(opt_level=3, config=config):
+                opt_mod, opt_params = relay.optimize(remainder_model, target=target, params=remainder_params)
```
Comment (masahi) [inline hummingbird/ml/_topology.py]: here too

File: hummingbird/ml/_topology.py (hunk 1 above)
Comment (interesaaat) [reply hummingbird/ml/_topology.py]: Ah, good to know! Let me fix it quickly.

**Review by masahi — APPROVED** [review]
we should definitely support dynamic batch size soon, to remove the need for `remainder_model stuff`

Excited this is finally getting merged!

File: .github/workflows/pythonapp.yml [hunk 3]
```diff
@@ -67,17 +64,70 @@ jobs:
...
       run: |
         # stop the build if there are Python syntax errors or undefined names
         flake8 . --count --select=E9,F63,F7,F82 --show-source --statistics
         # The GitHub editor is 127 chars wide
         flake8 . --count  --max-complexity=10 --max-line-length=127 --statistics
+    # We don't run pytest for Linux py3.7 since we do coverage for that case.
     - name: Test with pytest
-      run: |
-        pytest
+      if: ${{ matrix.python-version != '3.7' && matrix.os != 'ubuntu-latest' }}
```
Comment (ksaur) [inline .github/workflows/pythonapp.yml]: Looking at the [CI run](https://github.com/microsoft/hummingbird/actions/runs/338421930), it looks like ALL of the py3.7 runs (including macOS py3.7/windows py3.7) skip this `Test with pytest` phase.  

The logic _looks_ right, but maybe it is being parsed wrong.

File: hummingbird/ml/_container.py [hunk 4]
```diff
@@ -223,6 +217,48 @@ def score_samples(self, *inputs):
         return self.decision_function(*inputs) + self._extra_config[constants.OFFSET]
 
 
+# PyTorch containers.
```
Comment (ksaur) [inline hummingbird/ml/_container.py:236]: Next time it would be really helpful if a big refactoring like this was in a separate PR. It took me a while to understand how all of this finally relates to TVM with `class TVMSklearnContainer(SklearnContainer)` 

(Ex: do this refactoring, merge that, then add in the TVM stuff. It would be cleaner for all to understand looking back if we run into issues. :) )

File: hummingbird/ml/_topology.py [hunk 5]
```diff
@@ -185,6 +218,77 @@ def fix_graph(graph):
...
+                (
+                    topology.raw_model.input_names[i],
+                    remainder_trace_input[i].shape if type(remainder_trace_input) is tuple else remainder_trace_input.shape,
+                )
+                for i in range(len(topology.raw_model.input_names))
+            ]
+
+        # Pick the proper target.
+        if device == "cuda":
+            target = tvm.target.cuda()
```
Comment (ksaur) [inline hummingbird/ml/_topology.py:249]: i guess there is no way we can currently test this w/o GPU!

File: hummingbird/ml/_container.py [hunk 6]
```diff
@@ -136,28 +135,23 @@ def transform(self, *inputs):
...
         self._is_anomaly_detection = is_anomaly_detection
 
-    def _predict(self, *inputs):
-        if self._is_regression:
-            return self.model.forward(*inputs).cpu().numpy().ravel()
-        elif self._is_anomaly_detection:
-            return self.model.forward(*inputs)[0].cpu().numpy().ravel()
-        else:
-            return self.model.forward(*inputs)[0].cpu().numpy().ravel()
+    def _predict(self, *input):
```
Comment (ksaur) [inline hummingbird/ml/_container.py:158]: Can you please add  comments on the "pass" ones explaining the broader usage?

File: hummingbird/ml/_container.py [hunk 7]
```diff
@@ -478,16 +465,86 @@ def _decision_function(self, *inputs):
...
+    def _decision_function(self, *inputs):
+        if self._last_iteration and self._remainder_model is not None:
+            self._remainder_model.run(**self._to_tvm_tensor(*inputs))
+            scores = self._remainder_model.get_output(1).asnumpy().ravel()
+        else:
+            self.model.run(**self._to_tvm_tensor(*inputs))
+            scores = self.model.get_output(1).asnumpy().ravel()
+
+        # Backward compatibility for sklearn <= 0.21
+        if constants.IFOREST_THRESHOLD in self._extra_config:
```
Comment (ksaur) [inline hummingbird/ml/_container.py]: These 2 lines now appear in 3 places. consider to combine/make function?

File: hummingbird/ml/_utils.py [hunk 8]
```diff
@@ -113,6 +113,18 @@ def xgboost_installed():
...
 
+def tvm_installed():
+    """
+    Checks that *TVM* is available.
+    """
+    try:
+        import tvm
+        from tvm import relay
+    except ImportError:
+        return False
```
Comment (ksaur) [inline hummingbird/ml/_utils.py:124]: please add a test for this in `tests/test_no_extra_install.py`

File: hummingbird/ml/_topology.py [hunk 9]
```diff
@@ -40,20 +45,40 @@
...
     Utility function used to properly put the inputs into a format understandable by torch.
+    If a not None batch_size is passed, the function generates a tracing input of batch_size.
+    If input size % batch_size is not 0, a tracing input of the remainder is also generated.
     """
+    remainder = None
     if type(input) is tuple:
         if batch_size is not None:
             trace_input = tuple([torch.from_numpy(i)[0:batch_size, :] for i in input])
+            if len(input) > 0 and input[0].shape[0] % batch_size != 0:
+                remainder_size = input[0].shape[0] % batch_size
```
Comment (ksaur) [inline hummingbird/ml/_topology.py:68]: codecov says we have no tests for line 68-69. These seem like potential problem lines, add a test?

**Review by ksaur — COMMENTED** [review]
I'm halfway through, i need to look at topology a bit more and the tests but wanted to get you the first half now. :)

File: .github/workflows/pythonapp.yml (hunk 3 above)
Comment (interesaaat) [reply .github/workflows/pythonapp.yml]: Good catch! Let me try to rerun it.

Comment (interesaaat) [reply .github/workflows/pythonapp.yml]: Even if I rerun it is still skipping `pytest`. No idea what is going on.

File: hummingbird/ml/_topology.py (hunk 5 above)
Comment (interesaaat) [reply hummingbird/ml/_topology.py:249]: yea. It would be great to have GPU tests as well, but I don't think it is supported on Github Actions.

File: hummingbird/ml/_container.py (hunk 4 above)
Comment (interesaaat) [reply hummingbird/ml/_container.py:236]: I know...I just had to do this to simplify otherwise we were having a lot of duplicate code. I can pull it out and do a separate PR if you want.

File: hummingbird/ml/_container.py (hunk 6 above)
Comment (interesaaat) [reply hummingbird/ml/_container.py:158]: Done.

File: hummingbird/ml/_container.py (hunk 7 above)
Comment (interesaaat) [reply hummingbird/ml/_container.py]: I removed one occurrence, that was actually a bug.

File: hummingbird/ml/_container.py (hunk 4 above)
Comment (ksaur) [reply hummingbird/ml/_container.py:236]: No it's cool. I'll let this one slide haha :)

File: hummingbird/ml/_topology.py [hunk 10]
```diff
@@ -185,6 +218,77 @@ def fix_graph(graph):
...
+            ctx = tvm.cpu()
+        else:
+            raise RuntimeError("Device {} not recognized".format(device))
+
+        # Get configuration parameters.
+        config = {}
+        if constants.TVM_MAX_FUSE_DEPTH in extra_config:
+            config["relay.FuseOps.max_depth"] = extra_config[constants.TVM_MAX_FUSE_DEPTH]
+        else:
+            # 50 is a good depth for operator fusion. More than that will probably hurt performance.
```
Comment (ksaur) [inline hummingbird/ml/_topology.py:265]: If a depth of more than 50 will hit perf, should we print a warning on line 263, or have some sort of error checking there?  (not sure on this, just an idea)

File: hummingbird/ml/_topology.py [hunk 11]
```diff
@@ -185,6 +218,77 @@ def fix_graph(graph):
...
+            # https://github.com/microsoft/hummingbird/issues/232#issuecomment-697979508
+            config["relay.FuseOps.max_depth"] = 50
+
+        # Create the relay version of the model.
+        model, params = relay.frontend.from_pytorch(ts_model, test_input)
+        if remainder_trace_input is not None:
+            remainder_model, remainder_params = relay.frontend.from_pytorch(remainder_ts_model, remainder_test_input)
+
+        # Generate the model.
+        with tvm.transform.PassContext(opt_level=3, config=config):
```
Comment (ksaur) [inline hummingbird/ml/_topology.py:275]: why `opt_level=3`?  (is this a common knowledge, else add some comment?)

File: hummingbird/ml/operator_converters/_linear_implementations.py [hunk 12]
```diff
@@ -17,7 +17,7 @@ class LinearModel(BaseOperator, torch.nn.Module):
     def __init__(self, coefficients, intercepts, device, classes=[0], multi_class=None, is_linear_regression=False):
         super(LinearModel, self).__init__()
         self.coefficients = torch.nn.Parameter(torch.from_numpy(coefficients), requires_grad=False)
-        self.intercepts = torch.nn.Parameter(torch.from_numpy(intercepts), requires_grad=False)
+        self.intercepts = torch.nn.Parameter(torch.from_numpy(intercepts).view(-1), requires_grad=False)
```
Comment (ksaur) [inline hummingbird/ml/operator_converters/_linear_implementations.py:20]: why suddenly the `view(-1)`'s ?  Is this related to TVM? or the container changes?
or should this be yet a 3rd PR?  :-D

File: hummingbird/ml/operator_converters/sklearn/sv.py [hunk 13]
```diff
@@ -21,6 +21,7 @@ def __init__(self, kernel, degree, sv, nv, a, b, gamma, coef0, classes, device):
         super(SVC, self).__init__(classification=True)
         self.kernel = kernel
         self.degree = degree
+        # self.gamma = torch.nn.Parameter(torch.from_numpy(np.array(gamma, dtype=np.float64)), requires_grad=False)
```
Comment (ksaur) [inline hummingbird/ml/operator_converters/sklearn/sv.py]: why the commented-out line? should this be removed?

File: hummingbird/ml/operator_converters/sklearn/sv.py [hunk 14]
```diff
@@ -42,6 +43,7 @@ def __init__(self, kernel, degree, sv, nv, a, b, gamma, coef0, classes, device):
         if min(classes) != 0 or max(classes) != len(classes) - 1:
             self.perform_class_select = True
         self.n_classes = len(classes)
+        # self.two = torch.nn.Parameter(torch.from_numpy(np.array(2.0, dtype=np.float64)), requires_grad=False)
```
Comment (ksaur) [inline hummingbird/ml/operator_converters/sklearn/sv.py]: same question

File: hummingbird/ml/operator_converters/sklearn/sv.py [hunk 15]
```diff
@@ -52,7 +54,7 @@ def forward(self, x):
             # using quadratic expansion--susseptible to rounding-off errors
             # http://www.robots.ox.ac.uk/~albanie/notes/Euclidean_distance_trick.pdf
             x_norm = -self.gamma * (x ** 2).sum(1).view(-1, 1)
-            k = torch.exp(x_norm + self.sv_norm + 2.0 * self.gamma * torch.mm(x, self.sv_t))
+            k = torch.exp(x_norm + self.sv_norm + 2.0 * self.gamma * torch.mm(x, self.sv_t).double())
```
Comment (ksaur) [inline hummingbird/ml/operator_converters/sklearn/sv.py:55]: was this a bug? or why this change?

File: tests/test_onnxml_one_hot_encoder_converter.py [hunk 16]
```diff
@@ -51,6 +51,33 @@ def test_one_hot_encoder_onnx_int(self, rtol=1e-06, atol=1e-06):
         # Check that predicted values match
         np.testing.assert_allclose(onnx_ml_pred, onnx_pred, rtol=rtol, atol=atol)
 
+    # Test OneHotEncoder with 2 inputs
+    @unittest.skipIf(
+        not (onnx_ml_tools_installed() and onnx_runtime_installed()), reason="ONNXML test requires ONNX, ORT and ONNXMLTOOLS"
```
Comment (ksaur) [inline tests/test_onnxml_one_hot_encoder_converter.py:56]: What prompted this addition? I do not expect to see onnx-ml in this PR; am I missing something?

File: tests/test_sklearn_decision_tree_converter.py [hunk 17]
```diff
@@ -444,6 +484,224 @@ def test_decision_tree_ts_classifier_converter(self):
     def test_extra_trees_ts_classifier_converter(self):
         self._run_tree_classification_converter(ExtraTreesClassifier, 3, "torch.jit", n_estimators=10)
 
+    # Test trees with TVM backend
+    # Random forest gemm classifier
```
Comment (ksaur) [inline tests/test_sklearn_decision_tree_converter.py:488]: These test files have gotten rather long...  I'm not sure how many more we can hold! (ex: `skl->pyt, skl->ts, skl->tvm` all in the same file).  What file configuration do you think would be most straightforward for folks that might consider contributing to the project?  I know that trees are the longest. Maybe we are OK for now, but we should start planning for a refactor perhaps.

File: tests/test_sklearn_nb_converter.py [hunk 18]
```diff
@@ -100,18 +141,51 @@ def test_multinomialnb_classifer_multi_class_prior(self):
...
         X = np.random.rand(100, 200)
         X = np.array(X, dtype=np.float32)
         y = np.random.randint(num_classes, size=100) + labels_shift
 
         model.fit(X, y)
-        torch_model = hummingbird.ml.convert(model, "torch")
+        torch_model = hummingbird.ml.convert(model, backend, X)
         self.assertTrue(torch_model is not None)
-        np.testing.assert_allclose(model.predict_proba(X), torch_model.predict_proba(X), rtol=1e-6, atol=1e-5)
+        np.testing.assert_allclose(model.predict_proba(X), torch_model.predict_proba(X), rtol=1e-5, atol=1e-5)
```
Comment (ksaur) [inline tests/test_sklearn_nb_converter.py:188]: Is this the same problem as in #352 ?

File: tests/test_sklearn_normalizer_converter.py [hunk 19]
```diff
@@ -61,6 +63,46 @@ def test_float64_normalizer_converter(self):
                 model.transform(data), torch_model.transform(data_tensor), rtol=1e-06, atol=1e-06,
             )
 
+    # ONNX backend
+    @unittest.skipIf(not (onnx_runtime_installed()), reason="ONNXML test requires ONNX and  ORT")
```
Comment (ksaur) [inline tests/test_sklearn_normalizer_converter.py:67]: Was this onnx test moved from somewhere else?

Also,  I know that this is different from our [onnxml normalizer test](https://github.com/microsoft/hummingbird/blob/master/tests/test_onnxml_normalizer_converter.py#L27), but it is hella confusing.

Actually, at this point, even I am confused, and i wrote that.  Can you please add comments clarifying?

File: tests/test_sklearn_sv_converter.py [hunk 20]
```diff
@@ -126,6 +128,75 @@ def test_float64_linear_svc(self):
...
+    @unittest.skipIf(not (tvm_installed()), reason="TVM tests require TVM")
+    def test_svc_poly_tvm(self):
+        self._test_svc(2, kernel="poly", backend="tvm", extra_config={constants.TVM_MAX_FUSE_DEPTH: 30})
+
+    # NuSVC binary
+    @unittest.skipIf(not (tvm_installed()), reason="TVM tests require TVM")
+    def test_nu_svc_bi_tvm(self):
+        self._test_nu_svc(2, backend="tvm", extra_config={constants.TVM_MAX_FUSE_DEPTH: 30})
+
+    # # SVC multiclass
```
Comment (ksaur) [inline tests/test_sklearn_sv_converter.py:182]: future works?

File: tests/test_xgboost_converter.py [hunk 21]
```diff
@@ -263,6 +265,47 @@ def test_xgb_classifier_converter_torchscript(self):
...
+        for max_depth in [1, 3, 8, 10, 12]:
+            model = xgb.XGBRegressor(n_estimators=10, max_depth=max_depth)
+            np.random.seed(0)
+            X = np.random.rand(100, 200)
+            X = np.array(X, dtype=np.float32)
+            y = np.random.randint(1000, size=100)
+
+            model.fit(X, y)
+
+            torch_model = hummingbird.ml.convert(model, "torchscript", X, extra_config={constants.TVM_MAX_FUSE_DEPTH: 30})
```
Comment (ksaur) [inline tests/test_xgboost_converter.py]: did you mean `tvm` here not `torchscript`?

File: tests/test_xgboost_converter.py [hunk 22]
```diff
@@ -263,6 +265,47 @@ def test_xgb_classifier_converter_torchscript(self):
...
+        for max_depth in [1, 3, 8, 10, 12]:
+            model = xgb.XGBClassifier(n_estimators=10, max_depth=max_depth)
+            np.random.seed(0)
+            X = np.random.rand(100, 200)
+            X = np.array(X, dtype=np.float32)
+            y = np.random.randint(2, size=100)
+
+            model.fit(X, y)
+
+            torch_model = hummingbird.ml.convert(model, "tvm", X, extra_config={constants.TVM_MAX_FUSE_DEPTH: 30})
```
Comment (ksaur) [inline tests/test_xgboost_converter.py]: maybe `tvm_model` not `torch_model`?  same as line 285

File: tests/test_sklearn_scaler_converter.py [hunk 23]
```diff
@@ -107,6 +109,33 @@ def test_float64_robust_scaler_floats(self):
...
+    def test_standard_scaler_floats_tvm(self):
+        # Generate a random 2D array with values in [0, 1000)
+        np.random.seed(0)
+        data = np.random.rand(100, 200) * 1000
+        data = np.array(data, dtype=np.float32)
+        data_tensor = torch.from_numpy(data)
+
+        model = StandardScaler(with_mean=False, with_std=False)
+        model.fit(data)
+        torch_model = hummingbird.ml.convert(model, "tvm", data, extra_config={constants.TVM_MAX_FUSE_DEPTH: 30})
```
Comment (ksaur) [inline tests/test_sklearn_scaler_converter.py]: same comment as below, should these be not `torch_model`?

File: tests/test_sklearn_scaler_converter.py [hunk 24]
```diff
@@ -107,6 +109,33 @@ def test_float64_robust_scaler_floats(self):
...
+        self.assertIsNotNone(torch_model)
+        np.testing.assert_allclose(model.transform(data), torch_model.transform(data_tensor), rtol=1e-06, atol=1e-06)
+
+        model = StandardScaler(with_mean=True, with_std=False)
+        model.fit(data)
+        torch_model = hummingbird.ml.convert(model, "tvm", data, extra_config={constants.TVM_MAX_FUSE_DEPTH: 30})
+        self.assertIsNotNone(torch_model)
+        np.testing.assert_allclose(model.transform(data), torch_model.transform(data_tensor), rtol=1e-04, atol=1e-04)
```
Comment (ksaur) [inline tests/test_sklearn_scaler_converter.py]: did you mean this one to be `04` for `rtol/atol`?

File: tests/test_sklearn_scaler_converter.py [hunk 25]
```diff
@@ -107,6 +109,33 @@ def test_float64_robust_scaler_floats(self):
...
+        data = np.array(data, dtype=np.float32)
+        data_tensor = torch.from_numpy(data)
+
+        model = StandardScaler(with_mean=False, with_std=False)
+        model.fit(data)
+        torch_model = hummingbird.ml.convert(model, "tvm", data, extra_config={constants.TVM_MAX_FUSE_DEPTH: 30})
+        self.assertIsNotNone(torch_model)
+        np.testing.assert_allclose(model.transform(data), torch_model.transform(data_tensor), rtol=1e-06, atol=1e-06)
+
+        model = StandardScaler(with_mean=True, with_std=False)
```
Comment (ksaur) [inline tests/test_sklearn_scaler_converter.py]: this test seems formatted differently than others.  due to size?

**Review by ksaur — CHANGES_REQUESTED** [review]

File: hummingbird/ml/_utils.py (hunk 8 above)
Comment (interesaaat) [reply hummingbird/ml/_utils.py:124]: Ah, I forgot about this! Thanks! Added.

File: hummingbird/ml/_topology.py (hunk 10 above)
Comment (interesaaat) [reply hummingbird/ml/_topology.py:265]: The performance problem is model related. I think we already have a comment on the constant definition in supported. Do you believe we need to add also a warning on top of that?

File: hummingbird/ml/_topology.py (hunk 11 above)
Comment (interesaaat) [reply hummingbird/ml/_topology.py:275]: This enables all the TVM's optimizations. Added a comment.

File: hummingbird/ml/operator_converters/_linear_implementations.py (hunk 12 above)
Comment (interesaaat) [reply hummingbird/ml/operator_converters/_linear_implementations.py:20]: No no this is because of TVM. TVM is really sensible to tensor shapes.

File: hummingbird/ml/operator_converters/sklearn/sv.py (hunk 13 above)
Comment (interesaaat) [reply hummingbird/ml/operator_converters/sklearn/sv.py]: Ops, I think this is just some leftover from some test I was doing. I remove those.

File: hummingbird/ml/operator_converters/sklearn/sv.py (hunk 14 above)
Comment (interesaaat) [reply hummingbird/ml/operator_converters/sklearn/sv.py]: Same.

File: hummingbird/ml/operator_converters/sklearn/sv.py (hunk 15 above)
Comment (interesaaat) [reply hummingbird/ml/operator_converters/sklearn/sv.py:55]: This is because TVM wants the type right.

File: tests/test_onnxml_one_hot_encoder_converter.py (hunk 16 above)
Comment (interesaaat) [reply tests/test_onnxml_one_hot_encoder_converter.py:56]: I think I added this because a test using onnx-ml for one-hot encoding was missing. We were only directly converting to onnx using HB.

File: tests/test_sklearn_decision_tree_converter.py (hunk 17 above)
Comment (interesaaat) [reply tests/test_sklearn_decision_tree_converter.py:488]: I totally agree on this. Maybe we can split this file into decision_tree and random_forest? I think that going forward it would be good to have tests for all backends for each supported op.

File: tests/test_sklearn_nb_converter.py (hunk 18 above)
Comment (interesaaat) [reply tests/test_sklearn_nb_converter.py:188]: Yep.

File: tests/test_sklearn_sv_converter.py (hunk 20 above)
Comment (interesaaat) [reply tests/test_sklearn_sv_converter.py:182]: Yes, we are waiting for [some](https://github.com/pytorch/pytorch/issues/43867) operator in PyTorch.

File: tests/test_xgboost_converter.py (hunk 21 above)
Comment (interesaaat) [reply tests/test_xgboost_converter.py]: Ah yes! Good catch! Thanks.

File: tests/test_xgboost_converter.py (hunk 22 above)
Comment (interesaaat) [reply tests/test_xgboost_converter.py]: Yep.

File: tests/test_sklearn_scaler_converter.py (hunk 23 above)
Comment (interesaaat) [reply tests/test_sklearn_scaler_converter.py]: Yes, will fix all of them.

File: tests/test_sklearn_scaler_converter.py (hunk 25 above)
Comment (interesaaat) [reply tests/test_sklearn_scaler_converter.py]: mmm not sure, I see this formatted as the other tests.

File: tests/test_sklearn_scaler_converter.py (hunk 24 above)
Comment (interesaaat) [reply tests/test_sklearn_scaler_converter.py]: Nope. This one will actually fail if we increase the error. But this fails also in pytorch.I fixed it (for some reason we were using floats instead of doubles).

File: tests/test_sklearn_normalizer_converter.py (hunk 19 above)
Comment (interesaaat) [reply tests/test_sklearn_normalizer_converter.py:67]: No this is because we were missing a test for this operator for the onnx backend (without onnx-ml).

File: hummingbird/ml/_topology.py (hunk 9 above)
Comment (interesaaat) [reply hummingbird/ml/_topology.py:68]: Alright this was not supposed to happen. I have a small bug in the tests. Fixing it.

File: hummingbird/ml/_topology.py (hunk 10 above)
Comment (ksaur) [reply hummingbird/ml/_topology.py:265]: It think it's ok as-is. Seems more of a TVM issue than HB, so no warn seems fine

File: hummingbird/ml/operator_converters/_linear_implementations.py (hunk 12 above)
Comment (ksaur) [reply hummingbird/ml/operator_converters/_linear_implementations.py:20]: Interesting!  So..were these "wrong" before and we were just getting lucky?

File: tests/test_sklearn_decision_tree_converter.py (hunk 17 above)
Comment (ksaur) [reply tests/test_sklearn_decision_tree_converter.py:488]: Ah, splitting to "decsion_tree"/"random_forest" seems like an easy middle ground for now!  But  it's up to you...you could leave it for now and  we could do a more expansive test refactoring later

File: tests/test_sklearn_sv_converter.py (hunk 20 above)
Comment (ksaur) [reply tests/test_sklearn_sv_converter.py:182]: ah ok cool!  i know you have a comment somewhere else, but can you please link to that  here too?

File: tests/test_sklearn_scaler_converter.py (hunk 25 above)
Comment (ksaur) [reply tests/test_sklearn_scaler_converter.py]: Ah, what I meant was usually you'd have `def _test_standard_scaler_floats` and  then you'd call  that function 3 times rather than having 3 calls to `StandardScaler`.  But it's small enough that it's fine!  just an observation.

**Review by ksaur — APPROVED** [review]

File: hummingbird/ml/operator_converters/_linear_implementations.py (hunk 12 above)
Comment (interesaaat) [reply hummingbird/ml/operator_converters/_linear_implementations.py:20]: PyTorch is a bit more flexible regarding shapes and types. TVM is more compiler-style and requires shapes and types to perfectly match. I think that some operators where we still fail for TVM have the same kind of problem, so I hope they can be fixed quickly.

File: tests/test_sklearn_decision_tree_converter.py (hunk 17 above)
Comment (interesaaat) [reply tests/test_sklearn_decision_tree_converter.py:488]: Added issue #358.

File: tests/test_sklearn_normalizer_converter.py (hunk 19 above)
Comment (interesaaat) [reply tests/test_sklearn_normalizer_converter.py:67]: To add a little bit on this. The onnx tests (the one starting with the `onnx_` prefix) basically tests the onnx frontend (i.e., the onnx converters). But onnx can also be a backend, so this test makes sure that the path sklearn->pytorch->onnx works. At the end I hope we can get to a point where, for each operator, for each frontent, we a test for each different backend.

File: tests/test_sklearn_sv_converter.py (hunk 20 above)
Comment (interesaaat) [reply tests/test_sklearn_sv_converter.py:182]: yep. Done.

File: tests/test_sklearn_scaler_converter.py (hunk 25 above)
Comment (interesaaat) [reply tests/test_sklearn_scaler_converter.py]: Ah got it. Yes, let me fix this.

File: tests/test_sklearn_normalizer_converter.py (hunk 19 above)
Comment (ksaur) [reply tests/test_sklearn_normalizer_converter.py:67]: Ok, this should at least say "ONNX test" not `ONNXML test`

Comment (ksaur) [reply tests/test_sklearn_normalizer_converter.py:67]: Broader point....this distinction is clear to us, because we have been staring at "onnx" vs "onnx-ml" for a few years now. (for better or worse..... :) ) But ... the average person walking into this project will be very confused, especially since onnx-ml is severely under-documented!  Is there a way that we can explain this in a way that makes the project less intimidating to newcomers?   (not necessarily right here, but maybe in the documentation somewhere?)

Comment (interesaaat) [reply tests/test_sklearn_normalizer_converter.py:67]: Yes it should definitely say onnx test :)

I agree, we should explain this somewhere. Were you thinking to put this strictly in the documentation on the README or on a different document? Perhaps we can start to push some of this on the website.
